File: language-analyzer.go
package main

import (
	"context"
	"fmt"
	"io"
	"net"

	"google.golang.org/grpc"
	//"gopkg.in/bblfsh/client-go.v2/tools"
	"gopkg.in/bblfsh/client-go.v2/tools"
	"gopkg.in/src-d/go-log.v1"
	"gopkg.in/src-d/lookout-sdk.v0/pb"
)

//Example Analyser gRPC service implementation.
//Posts file-level comments for every file with language detected.

type analyzer struct{}

var portToListen = 2020
var dataSrvAddr = "localhost:10301"
var version = "alpha"

//TODO(bzz): max msg size

func (*analyzer) NotifyReviewEvent(ctx context.Context, review *pb.ReviewEvent) (*pb.EventResponse, error) {
	log.Infof("got review request %v", review)

	conn, err := grpc.Dial(dataSrvAddr, grpc.WithInsecure())
	if err != nil {
		log.Errorf(err, "failed to connect to DataServer at %s", dataSrvAddr)
	}
	defer conn.Close()

	dataClient := pb.NewDataClient(conn)
	changes, err := dataClient.GetChanges(ctx, &pb.ChangesRequest{
		Head:            &review.Head,
		Base:            &review.Base,
		WantContents:    false,
		WantUAST:        true,
		ExcludeVendored: true,
	})
	if err != nil {
		log.Errorf(
			err, "GetChanges from DataServer %s failed", dataSrvAddr)
	}

	var comments []*pb.Comment
	for {
		change, err := changes.Recv()
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Errorf(err, "GetChanges from DataServer %s failed", dataSrvAddr)
		}

		log.Infof("analyzing '%s' in %s", change.Head.Path, change.Head.Language)
		//TODO: put your analysis here!
		query := "//*[@roleFunction]"
		fns, err := tools.Filter(change.Head.UAST, query)
		if err != nil {
			log.Errorf(err, "quering UAST from %s with %s failed", change.Head.Path, query)
		}

		comments = append(comments, &pb.Comment{
			File: change.Head.Path,
			Line: 0,
			Text: fmt.Sprintf("language: %s, functions: %d", change.Head.Language, len(fns)),
		})
	}

	return &pb.EventResponse{AnalyzerVersion: version, Comments: comments}, nil
}

func (*analyzer) NotifyPushEvent(context.Context, *pb.PushEvent) (*pb.EventResponse, error) {
	return &pb.EventResponse{}, nil
}

func main() {
	lis, err := net.Listen("tcp", fmt.Sprintf("0.0.0.0:%d", portToListen))
	if err != nil {
		log.Errorf(err, "failed to listen on port: %d", portToListen)
	}

	grpcServer := grpc.NewServer()
	pb.RegisterAnalyzerServer(grpcServer, &analyzer{})
	log.Infof("starting gRPC Analyzer server at port %d", portToListen)
	grpcServer.Serve(lis)
}

File: language-analyzer.py
#!/usr/bin/env python3

"""
 Example Analyzer gRPC service implementation.
 Posts file-level comments for every file with language detected.
"""

from concurrent.futures import ThreadPoolExecutor

import time
import grpc
from lookout.sdk import pb
from lookout.sdk.grpc import to_grpc_address, create_channel

from bblfsh import filter as filter_uast

port_to_listen = 2021
data_srv_addr = to_grpc_address("ipv4://localhost:10301")
version = "alpha"


class Analyzer(pb.AnalyzerServicer):
    def NotifyReviewEvent(self, request, context):
        print("got review request {}".format(request))

        # client connection to DataServe
        channel = create_channel(data_srv_addr)
        stub = pb.DataStub(channel)
        changes = stub.GetChanges(
            pb.ChangesRequest(
                head=request.commit_revision.head,
                base=request.commit_revision.base,
                want_contents=False,
                want_uast=True,
                exclude_vendored=True))

        comments = []
        for change in changes:
            if not change.HasField("head"):
                continue

            print("analyzing '{}' in {}".format(
                change.head.path, change.head.language))
            fns = list(filter_uast(change.head.uast, "//*[@roleFunction]"))
            comments.append(
                pb.Comment(
                    file=change.head.path,
                    line=0,
                    text="language: {}, functions: {}".format(change.head.language, len(fns))))
        return pb.EventResponse(analyzer_version=version, comments=comments)

    def NotifyPushEvent(self, request, context):
        pass


def serve():
    server = grpc.server(thread_pool=ThreadPoolExecutor(max_workers=10))
    pb.add_analyzer_to_server(Analyzer(), server)
    server.add_insecure_port("0.0.0.0:{}".format(port_to_listen))
    server.start()

    one_day_sec = 60 * 60 * 24
    try:
        while True:
            time.sleep(one_day_sec)
    except KeyboardInterrupt:
        server.stop(0)


def main():
    print("starting gRPC Analyzer server at port {}".format(port_to_listen))
    serve()


if __name__ == "__main__":
    main()
